src/js/module.js

```diff
@@ -170,54 +170,77 @@ iotjs_module_t.tryPath = function(path) {
-    return Native.require(id);
-  }
-  var module = new iotjs_module_t(id, parent);
+iotjs_module_t.load = function(id, parent, isRemote) {
+  if (isRemote) {
+    if (process.builtin_modules[id[0]]) {
+      return Native.require(id);
+    }
 
-  var modPath = iotjs_module_t.resolveModPath(module.id, module.parent);
+    var module = new iotjs_module_t(id[0], parent);
+    var cachedModule = iotjs_module_t.cache[id[0]];
 
-  var cachedModule = iotjs_module_t.cache[modPath];
-  if (cachedModule) {
-    return cachedModule.exports;
-  }
+    if (cachedModule) {
+      return cachedModule.exports;
+    }
 
-  if (!modPath) {
-    throw new Error('Module not found: ' + id);
-  }
+    module.filename = id[0];
+    module.compile(true, id[0], id[1]);
+    iotjs_module_t.cache[id[0]] = module;
 
-  module.filename = modPath;
-  module.dirs = [modPath.substring(0, modPath.lastIndexOf('/') + 1)];
+    return module.exports;
+  } else {
+    if (process.builtin_modules[id]) {
+      return Native.require(id);
+    }
+    var module = new iotjs_module_t(id, parent);
+    var modPath = iotjs_module_t.resolveModPath(module.id, module.parent);
+    var cachedModule = iotjs_module_t.cache[modPath];
 
-  var ext = modPath.substr(modPath.lastIndexOf('.') + 1);
+    if (cachedModule) {
+      return cachedModule.exports;
+    }
 
-  if (ext === 'js') {
-    module.compile();
-  } else if (ext === 'json') {
-    var source = process.readSource(modPath);
-    module.exports = JSON.parse(source);
-  }
+    if (!modPath) {
+      throw new Error('Module not found: ' + id);
+    }
+
+    module.filename = modPath;
+    module.dirs = [modPath.substring(0, modPath.lastIndexOf('/') + 1)];
+
+    var ext = modPath.substr(modPath.lastIndexOf('.') + 1);
+
+    if (ext === 'js') {
+      module.compile(false, null, null);
+    } else if (ext === 'json') {
+      var source = process.readSource(modPath);
+      module.exports = JSON.parse(source);
+    }
 
-  iotjs_module_t.cache[modPath] = module;
+    iotjs_module_t.cache[modPath] = module;
 
-  return module.exports;
+    return module.exports;
+  }
 };
 
 
-iotjs_module_t.prototype.compile = function() {
-  var source = process.readSource(this.filename);
-  var fn = process.compile(this.filename, source);
-  fn.call(this.exports, this.exports, this.require.bind(this), this);
+iotjs_module_t.prototype.compile = function(remote, _filename, _source) {
+  if (!remote) {
+    var source = process.readSource(this.filename);
+    var fn = process.compile(this.filename, source);
+    fn.call(this.exports, this.exports, this.require.bind(this), this);
+  } else {
+    var fn = process.compile(_filename, _source);
+    fn.call(this.exports, this.exports, this.require.bind(this), this);
+  }
 };
 
 
 iotjs_module_t.runMain = function() {
   if (process.debuggerWaitSource) {
-    var fn = process.debuggerSourceCompile();
-    fn.call();
+    var fn = process.getRemoteSource();
+    iotjs_module_t.load(fn, null, true);
```

Comment:
I think it would be better to have a separate function for "remote load" so then we don't need to pass around the third parameter.